### PR TITLE
Swift: add SetContent for data flow

### PR DESCRIPTION
**Pull request by rdmarsh2**

### CODE REVIEW

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Set.qll [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+/**
+ * A model for `Set` and related class members that permit data flow.
+ */
+private class SetSummaries extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        ";Set;true;insert(_:);;;Argument[-1].SetElement;ReturnValue.TupleElement[1];value",
+        ";Set;true;insert(_:);;;Argument[0];Argument[-1].SetElement;value",
```
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Set.qll:11]: Should there be value flow from `Argument[0]` to `ReturnValue.TupleElement[1]` as well?

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Set.qll [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+        ";Set;true;randomElement();;;Argument[-1].SetElement;ReturnValue;value"
```
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Set.qll:12]: I think this should be to `ReturnValue.OptionalSome`, now that we model enum content.

File: swift/ql/test/library-tests/dataflow/dataflow/test.swift [hunk 3]
```diff
@@ -662,3 +662,44 @@ func testSwap() {
...
+}
+
+func testSetCollections() {
+    var set1: Set = [1,2,3]
+    sink(arg: set1.randomElement()!)
+    set1.insert(source())
+    sink(arg: set1.randomElement()!) // $ MISSING: flow=700
+
+    let set2 = Set([source()])
+    sink(arg: set2.randomElement()!) // $ MISSING: flow=703
```
Comment (geoffw0) [inline swift/ql/test/library-tests/dataflow/dataflow/test.swift:704]: I think you'll need a model of the `Set` initializer to get this case working?

**Comment by MathiasVP** [comment]
Also small nitpick: How about just calling this `CollectionContent`? There's no reason why this `Content` can't be extended for non-`Set` things in a future PR.

**Comment by geoffw0** [comment]
> How about just calling this CollectionContent?

I like this idea.

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Set.qll (hunk 1 above)
Comment (rdmarsh2) [reply swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Set.qll:11]: Yes, there should

File: swift/ql/test/library-tests/dataflow/dataflow/test.swift (hunk 3 above)
Comment (geoffw0) [reply swift/ql/test/library-tests/dataflow/dataflow/test.swift:704]: I see this has been added. :+1:

File: swift/ql/lib/codeql/swift/dataflow/internal/FlowSummaryImplSpecific.qll [hunk 4]
```diff
@@ -110,6 +110,11 @@ private string getContentSpecific(ContentSet cs) {
     result = "Field[" + c.getField().getName() + "]"
   )
   or
+  exists(Content::EnumContent c |
+    cs.isSingleton(c) and
+    result = "EnumElement[" +  c.getSignature() + "]"
+  )
+  or
```
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/dataflow/internal/FlowSummaryImplSpecific.qll:117]: Thanks for adding this, I missed it in https://github.com/github/codeql/pull/13795 .

File: swift/ql/test/library-tests/dataflow/dataflow/test.swift [hunk 5]
```diff
@@ -692,14 +692,17 @@ func testArray() {
...
 func testSetCollections() {
     var set1: Set = [1,2,3]
     sink(arg: set1.randomElement()!)
     set1.insert(source())
-    sink(arg: set1.randomElement()!) // $ MISSING: flow=700
+    sink(arg: set1.randomElement()!) // $flow=703
 
     let set2 = Set([source()])
-    sink(arg: set2.randomElement()!) // $ MISSING: flow=703
+    sink(arg: set2.randomElement()!) // $ flow=706
```
Comment (geoffw0) [inline swift/ql/test/library-tests/dataflow/dataflow/test.swift:707]: 🎉

**Comment by MathiasVP** [comment]
Do you mind rebasing this on the latest main so that we get a smaller diff @rdmarsh2?

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll [hunk 6]
```diff
@@ -255,7 +255,8 @@ private module Cached {
     TFieldContent(FieldDecl f) or
     TTupleContent(int index) { exists(any(TupleExpr te).getElement(index)) } or
     TEnumContent(ParamDecl f) { exists(EnumElementDecl d | d.getAParam() = f) } or
-    TArrayContent()
+    TArrayContent() or
+    TSetContent()
```
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:260]: I still think we should call this `CollectionContent` (see https://github.com/github/codeql/pull/13838#issuecomment-1658303515)

File: swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Set.qll [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+ * A model for `Set` and related class members that permit data flow.
+ */
+private class SetSummaries extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        ";Set;true;insert(_:);;;Argument[-1].SetElement;ReturnValue.TupleElement[1];value",
+        ";Set;true;insert(_:);;;Argument[0];Argument[-1].SetElement;value",
+        ";Set;true;insert(_:);;;Argument[0];ReturnValue.TupleElement[1];value",
+        ";Set;true;init(_:);;;Argument[0].ArrayElement;ReturnValue.SetElement;value"
```
Comment (geoffw0) [inline swift/ql/lib/codeql/swift/frameworks/StandardLibrary/Set.qll:17]:
```suggestion
        ";Set;true;insert(_:);;;Argument[-1].CollectionElement;ReturnValue.TupleElement[1];value",
        ";Set;true;insert(_:);;;Argument[0];Argument[-1].CollectionElement;value",
        ";Set;true;insert(_:);;;Argument[0];ReturnValue.TupleElement[1];value",
        ";Set;true;init(_:);;;Argument[0].ArrayElement;ReturnValue.CollectionElement;value"
```

**Review by MathiasVP — DISMISSED** [review]
LGTM once we have a successful DCA run!

**Review by MathiasVP — APPROVED** [review]

**Review by geoffw0 — APPROVED** [review]
This looks great, and I could use the `readStep` fix in two of my open PRs and `CollectionContent` in one, so the sooner we merge this the better!

**Comment by geoffw0** [comment]
DCA LGTM.  Merging.

**Pull request merged**